File: python/lcm_p0151_reverse_words_in_a_string.py
"""
LCM 151. Reverse Words in a String

Given a string s, reverse only all the vowels in the string and return it.

The vowels are 'a', 'e', 'i', 'o', and 'u', and they can appear in both lower and upper cases, more than once.

Constraints:
- 1 <= s.length <= 3 * 105
- s consist of printable ASCII characters.

Topics:
- Two Pointers
- String
"""


class Solution:

    # Time Complexity: O(n) - 0 ms -> 100.00%
    # Space Complexity: O(1) - 18.07 MB -> 24.29%
    def reverseWords(self, s: str) -> str:
        words = s.split()
        return " ".join(words[i] for i in range(len(words) - 1, -1, -1))

# split() in Python treats consecutive whitespaces as a single separator and removes leading / trailing whitespaces

File: java/lcm_p0151_reverse_words_in_a_string.java
/*
  LCM 151. Reverse Words in a String

  Given an input string s, reverse the order of the words.

  A word is defined as a sequence of non-space characters.
  The words in s will be separated by at least one space.

  Return a string of the words in reverse order concatenated by a single space.

  Note that s may contain leading or trailing spaces or multiple spaces between two words.
  The returned string should only have a single space separating the words. Do not include any extra spaces.

  Constraints:
  - 1 <= s.length <= 10^4
  - s contains English letters (upper-case and lower-case), digits, and spaces ' '.
  - There is at least one word in s.

  Topics: Two Pointers, String
*/

class Solution {
  public String reverseWords(String s) {

    // note: "\\s+" matches sequence of one or more whitespace characters.
    String[] words = s.trim().split("\\s+");

    reverseArray(words);

    return String.join(" ", words);
  }

  // alternative: use Collections.reverse(List<T>) instead.
  private void reverseArray(String[] words) {

    int n = words.length;

    for (int i = 0; i < n / 2; i++) {
      String temp = words[i];
      words[i] = words[n - i - 1];
      words[n - i - 1] = temp;
    }
  }
}

// Time Complexity: O(L) - 8 ms -> 39.40%
// Space Complexity: O(L) - 42.65 MB -> 94.35%
// L => length of string s